File: branches/PetersOrcMachine/OrcScala/src/orc/lib/math/Sub.java
//
// Sub.java -- Java class Sub
// Project OrcScala
//
// $Id$
//
// Copyright (c) 2009 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.math;

import java.math.BigDecimal;
import java.math.BigInteger;

import orc.error.runtime.TokenException;
import orc.types.Type;
import orc.values.sites.TypedSite;
import orc.values.sites.compatibility.Args;
import orc.values.sites.compatibility.Args.NumericBinaryOperator;
import orc.values.sites.compatibility.EvalSite;
import orc.values.sites.compatibility.Types;

public class Sub extends EvalSite implements TypedSite {
	@SuppressWarnings("synthetic-access")
	private static final MyOperator op = new MyOperator();

	private static final class MyOperator implements NumericBinaryOperator<Number> {
		@Override
		public Number apply(final BigInteger a, final BigInteger b) {
			return a.subtract(b);
		}

		@Override
		public Number apply(final BigDecimal a, final BigDecimal b) {
			return a.subtract(b);
		}

		@Override
		public Number apply(final int a, final int b) {
			return Integer.valueOf(a - b);
		}

		@Override
		public Number apply(final long a, final long b) {
			return Long.valueOf(a - b);
		}

		@Override
		public Number apply(final byte a, final byte b) {
			return Integer.valueOf(a - b);
		}

		@Override
		public Number apply(final short a, final short b) {
			return Integer.valueOf(a - b);
		}

		@Override
		public Number apply(final double a, final double b) {
			return Double.valueOf(a - b);
		}

		@Override
		public Number apply(final float a, final float b) {
			return Float.valueOf(a - b);
		}
	}

	@Override
	public Object evaluate(final Args args) throws TokenException {
		return Args.applyNumericOperator(args.numberArg(0), args.numberArg(1), op);
	}

	@Override
	public Type orcType() {
		return Types.overload(Types.function(Types.integer(), Types.integer(), Types.integer()), Types.function(Types.number(), Types.number(), Types.number()));
	}
}

File: OrcScala/src/orc/lib/math/Sub.scala
//
// Sub.scala -- Scala object Sub
// Project OrcScala
//
// Created by amp on Sept 25, 2017.
//
// Copyright (c) 2018 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.math

import orc.values.sites.{ FunctionalSite, IllegalArgumentInvoker, LocalSingletonSite, OverloadedDirectInvokerMethod2 }

object Sub extends OverloadedDirectInvokerMethod2[Number, Number] with FunctionalSite with Serializable with LocalSingletonSite {
  def getInvokerSpecialized(arg1: Number, arg2: Number) = {
    // TODO: This does not handle all possible reasonable cases and some of the priorities are weird. When we improve the numeric stack we should fix this.
    (arg1, arg2) match {
      case (a: BigDecimal, b: BigDecimal) =>
        invoker(a, b)(_ - _)
      case (a: java.lang.Double, b: java.lang.Double) =>
        invokerInline(a, b)(_.doubleValue() - _.doubleValue())
      case (a: java.math.BigDecimal, b: java.math.BigDecimal) =>
        invoker(a, b)(_.subtract(_))
      case (a: java.lang.Long, b: java.lang.Long) =>
        invokerInline(a, b)(_.longValue() - _.longValue())
      case (a: java.lang.Integer, b: java.lang.Integer) =>
        invokerInline(a, b)(_.intValue() - _.intValue())
      case (a: BigInt, b: BigInt) =>
        invoker(a, b)(_ - _)
      case (a: java.math.BigInteger, b: java.math.BigInteger) =>
        invoker(a, b)(_.subtract(_))
      case (a: BigDecimal, b: Number) =>
        invoker(a, b)(_ - _.doubleValue())
      case (a: Number, b: BigDecimal) =>
        invoker(a, b)(_.doubleValue() - _)
      case (a: java.lang.Double, b: Number) =>
        invoker(a, b)(_.doubleValue() - _.doubleValue())
      case (a: Number, b: java.lang.Double) =>
        invoker(a, b)(_.doubleValue() - _.doubleValue())
      case (a: BigInt, b: Number) =>
        invoker(a, b)(_ - _.longValue())
      case (a: Number, b: BigInt) =>
        invoker(a, b)(_.longValue() - _)
      case (a: java.lang.Long, b: Number) =>
        invoker(a, b)(_.longValue() - _.longValue())
      case (a: Number, b: java.lang.Long) =>
        invoker(a, b)(_.longValue() - _.longValue())
      case (a: Number, b: Number) =>
        IllegalArgumentInvoker(this, Array(a, b))
    }
  }
  override def toString = "Sub"
}
